src/validator.cpp

```diff
@@ -1635,6 +1708,530 @@ void Validator::ValidatorImpl::validateMathMLElements(const XmlNodePtr &node, co
+                           component);
+
+            return;
+        }
+
+        auto cnType = node->attribute("type");
+
+        if (cnType.empty() || (cnType == "real")) {
+            auto ok = (nonCommentChildCount(node) != 1) ? false : nonCommentChildNode(node, 0)->isBasicReal();
+
+            if (!ok) {
+                addMathMLIssue("Math has a 'cn' element of 'real' type with no valid text node (representing a basic number) as a child.",
+                               Issue::ReferenceRule::MATH_CN_FORMAT,
+                               component);
+            }
+        } else if (cnType == "e-notation") {
+            auto ok = false;
+
+            if (nonCommentChildCount(node) == 3) {
+                ok = nonCommentChildNode(node, 0)->isBasicReal()
+                     && nonCommentChildNode(node, 1)->isMathmlElement("sep")
+                     && nonCommentChildNode(node, 2)->isInteger();
+            }
+
+            if (!ok) {
+                addMathMLIssue("Math has a 'cn' element of 'e-notation' type with no valid text node (representing a basic number), no 'sep' element, and/or no valid text node (representing an integer) as children.",
+                               Issue::ReferenceRule::MATH_CN_FORMAT,
+                               component);
+            }
+        } else {
+            addMathMLIssue("Math has a 'cn' element which is not of 'real' or 'e-notation' type.",
+                           Issue::ReferenceRule::MATH_CN_FORMAT,
+                           component);
+        }
+
+        // Qualifier elements.
+
+    } else if (node->isMathmlElement("degree")) {
+        // A 'degree' element can be used either with a 'root' element or within
+        // a 'bvar' element, e.g.
+        //
+        //   <apply>
+        //     <diff/>
+        //     <bvar>
+        //       <ci>t</ci>
+        //       <degree>
+        //         <cn cellml:units="dimensionless">2</cn>
+        //       </degree>
+        //     </bvar>
+        //     <ci>x</ci>
+        //   </apply>
+        //
+        // and
+        //
+        //   <apply>
+        //     <root/>
+        //     <degree>
+        //       <cn cellml:units="dimensionless">3</cn>
+        //     </degree>
+        //     <ci>a</ci>
+        //   </apply>
+
+        auto parentNode = node->parent();
+        auto siblingCount = hasOneOrTwoMathmlSiblings(parentNode, node, component);
+
+        if (siblingCount == 1) {
+            // Used with a 'degree' element.
+
+            isSecondMathmlChild(parentNode, node, component)
+                && hasOneMathmlChild(node, component);
+        } else if (siblingCount == 2) {
+            // Used with a 'root' element.
+
+            hasFirstMathmlSiblingWithName(parentNode, node, "root", component)
+                && isSecondMathmlChild(parentNode, node, component)
+                && hasOneMathmlChild(node, component);
+        }
+    } else if (node->isMathmlElement("logbase")) {
```

Comment:
as for degree, this duplicates the checks done for `log`...

Reply:
No, it doesn't. Someone could perfectly use `logbase` without using `log` and this is what those checks are about. See [this CellML file](https://github.com/agarny/libcellml/blob/issue1082/tests/resources/invalidmathmlelementschildrenorsiblings.cellml) and search for `logbase` to see what I mean.

Reply:
doh! you're right :)